dev/ci/ci-brick.sh

```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+
+set -e
+
+ci_dir="$(dirname "$0")"
+. "${ci_dir}/ci-common.sh"
+
+git_download brick
+
+# Extract required version of Iris (avoiding "+" which does not work on MacOS :( *)
```

Comment:
Ths won't work properly with `ci-brick: ci-iris` in the makefile, `make ci-brick` will download and build iris with the version from iris-examples then the `git_download iris` in this script will be a noop.

I am also quite reluctant to add more jobs using this dependency management as it is a pain with overlays.

Reply:
I don't mind doing things differently but I am not sure what the alternative is.

Reply:
If you're not willing to commit to keeping master branches compatible with each other the alternative would be to have for-coq-ci branches that get updated to co-compatible commits, I guess at the same time that you update the .opam or whenever is convenient 
The overlays would be merged into for-coq-ci and backported to master when convenient

Reply:
We have so many dependencies that we already know that we cannot keep BRiCk's coq-ci branch up-to-date with our own master. So we'll do sporadic pushes whenever we have time to port to new versions of coq and new versions of our dependencies. I guess I was trying to keep these discrepancies to a minimum but a big practical part of it was that iris and stdpp don't have `dev` versions in opam. I can try to pin them to `master` for now.

My worry is that BRiCk could break Coq CI simply by pulling in an incompatible stdpp master and iris master. iris master is checked against stdpp master nightly AFAICT but somebody needs to come in and fix the breakage so there could be long-ish delays. (I suspect in practice it's a few days at most but still..)

Also, it seems weird that BRiCk would be the only thing testing iris and stdpp master. They don't seem to be tested on their own against coq master. Or am I missing something?

Reply:
The idea would be that iris and stdpp also get for-ci branches

Reply:
I see. That's something that stdpp and iris developers would have to orchestrate, I think.

Reply:
Maybe we should even encourage Iris developers to add some reverse-dependency compatibility testing, like MathComp has. It's really difficult to keep testing compatibility of Coq dev with projects that depend on something other than Iris dev (or whatever version Coq tests and Coq developers can send overlays for).

Reply:
The overlay manager solution we discussed at the CUDW would help with that I think, as it precisely designed to solve (in combination with Dune) these kind of consistency problems.

While that's not ready yet, if there is interest we could work on it.

Reply:
We do have reverse-dep compat checking in Iris. However, it is on-demand: we know we are regularly landing breaking changes in Iris master, and we don't want to block landing Iris changes on everyone porting their things. (In fact that wouldn't even work: the fixed reverse-dep would fail to build against older Iris! We'd have to merge the updates in both atomically.) We expect users of Iris will use a pinned dependency and bump that when they have the time.

I don't see how to practically have a for-ci Iris branch that works for examples, perennial, and brick. Those developments will not usually get updated for breaking Iris changes in lock-step.

We could in principle merge iris/examples into the main Iris repo (we have "megarepo" support nowadays) and therefore ensure that it gets fixed atomically with Iris changes. But that won't help for perennial and brick.

Reply:
What's the cost for Coq devs, more specifically? Must you prepare similar overlays for different Iris snapshots, or something else?

> I don't see how to practically have a for-ci Iris branch that works for examples, perennial, and brick. Those developments will not usually get updated for breaking Iris changes in lock-step.

That branch doesn't have to track master closely. If you pick a common snapshot (like the latest release, for example, but something more recent works), then the Coq devs and contributors must only port that version of Iris to Coq master. IIRC I've seen this idea in other Coq-CI projects, and that seems also what @Janno described?

But that seems to save work for Coq devs at some (smaller?) cost for Iris/Perennial devs (merging overlays into their master) — that's for you to negotiate.

Also, Iris users would need to coordinate on "sync" points when for-ci pulls master.

Reply:
>What's the cost for Coq devs, more specifically? Must you prepare similar overlays for different Iris snapshots, or something else?

I think so, but currently it is simply broken if iris-examples and brick want incompatible iris versions

Reply:
Overall maybe putting projects with no backwards incompatibility with their dependents in coq CI is not a great idea.

Reply:
> If you pick a common snapshot (like the latest release, for example, but something more recent works), then the Coq devs and contributors must only port that version of Iris to Coq master. 

If it receives patches to be compatible it's not a snapshot any more. ;)

> Overall maybe putting projects with no backwards incompatibility with their dependents in coq CI is not a great idea.

There is basically nothing one can do with a Coq library, except for adding new things, that would be backwards compatible. (And even then, adding new hints can cause loops.) You can't ever change any definition. For Iris, that is definitely not a realistic option. We need to be able to evolve the library.

---

So far, the way to deal with Iris breakages wasn't snapshots, it was sending us an MR that we land. Then examples and Perennial have to be updated (which the respective devs will take care of), and then CI should work again. With BRiCk now that would mean that the BRiCk devs also have to update to latest Iris whenever that happens. I don't think Coq devs need to do more overlays than before?

Reply:
>So far, the way to deal with Iris breakages wasn't snapshots, it was sending us an MR that we land.

and it is the hardest process of all the projects in CI IMO, which is why I don't want it to expand

Reply:
I'm open for suggestions for how we could improve this. However, I don't think it becomes *worse* with adding BRiCk (if the BRiCk devs commit to quickly updating things to latest Iris): all Coq devs need to do is make an Iris MR.